src/daemon/backends/connman_agent.cpp

```diff
@@ -0,0 +1,336 @@
+// Copyright (C) 2019 Luxoft Sweden AB
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#include "daemon/backends/connman_agent.h"
+
+#include <giomm.h>
+#include <glibmm.h>
+
+#include <cstdint>
+#include <optional>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+
+#include "common/credentials.h"
+#include "common/dbus.h"
+#include "common/string_to_valid_utf8.h"
+
+namespace ConnectivityManager::Daemon
+{
+    namespace
+    {
+        constexpr char FIELD_NAME_HIDDEN_SSID_UTF8[] = "Name";
+        constexpr char FIELD_NAME_HIDDEN_SSID[] = "SSID";
+        constexpr char FIELD_NAME_EAP_USERNAME[] = "Identity";
+        constexpr char FIELD_NAME_PASSPHRASE[] = "Passphrase";
+        constexpr char FIELD_NAME_PREVIOUS_PASSPHRASE[] = "PreviousPassphrase";
+        constexpr char FIELD_NAME_WPS[] = "WPS";
+        constexpr char FIELD_NAME_WISPR_USERNAME[] = "Username";
+        constexpr char FIELD_NAME_WISPR_PASSWORD[] = "Password";
+
+        constexpr char FIELD_ARGUMENT_TYPE[] = "Type";
+        constexpr char FIELD_ARGUMENT_VALUE[] = "Value";
+        // constexpr char FIELD_ARGUMENT_REQUIREMENT[] = "Requirement"; Not checked at the moment.
+        // constexpr char FIELD_ARGUMENT_ALTERNATES[] = "Alternates"; Not checked at the moment.
+
+        constexpr char FIELD_ARGUMENT_TYPE_PSK[] = "psk";
+        constexpr char FIELD_ARGUMENT_TYPE_WEP[] = "wep";
+        constexpr char FIELD_ARGUMENT_TYPE_PASSPHRASE[] = "passphrase";
+        constexpr char FIELD_ARGUMENT_TYPE_RESPONSE[] = "response";
+        constexpr char FIELD_ARGUMENT_TYPE_WPS_PIN[] = "wpspin";
+        constexpr char FIELD_ARGUMENT_TYPE_STRING[] = "string";
+        // constexpr char FIELD_ARGUMENT_TYPE_SSID[] = "ssid"; Not checked at the moment.
+
+        using Fields = std::map<Glib::ustring, Glib::VariantBase>;
+        using Credentials = Common::Credentials;
+        using Password = Common::Credentials::Password;
+
+        std::optional<Password> value_to_password(const Glib::ustring &name,
+                                                  const Glib::VariantBase &variant)
+        {
+            // FIXME: Have not tested having Glib::ustring as value here. More convenient but might
```

Comment:
Forgot this FIXME... will fix. (Used FIXME for "must fix" and TODO for "punt" for now.)